server/src/main.c

```diff
@@ -0,0 +1,404 @@
+ * @param server_addrinfo pointer to pointer of the servers addrinfo
+ * @return 0 if successful
+ * @return 1 if parent process that needs to be ended
+ * @return -1 otherwise
+ */
+static int setup_socket_server(const bool execute_as_daemon, int *server_socket,
+                               struct addrinfo **server_addrinfo);
+
+/**
+ * @brief Sets up the SIGINT, SIGTERM, and SIGUSR1 signal handlers.
+ * @return 0 if successful
+ * @return 1 otherwise
+ */
+static int setup_signal_handlers(timer_t *sugusr1_timer);
+
+/**
+ * @brief Intended to be run in a thread. Completes the data transfer from the
+ * client passed in `arg`. Data from the client is recieved and written to file,
+ * then the entire contents of the file are sent back to the client.
+ * @param arg client file descriptor as an int*
+ */
+static void *data_transfer_worker(void *arg);
+
+/**
+ * @brief Writes a timestamp in RFC 2822 complient format to `RESULT_FILE`
+ * triggered by the `timestamp_sem`. This is inteneded to be run in a thread.
+ */
+static void *log_timestamp_worker(void *arg);
+
+/**
+ * @brief SIGINT signal handler used to gracefully shutdown the program
+ * @param sig signal number
+ */
+static void sigint_handler(int sig);
+
+/**
+ * @brief SIGUSR1 signal handler used to handle timestamp logging events
+ * @param sig signal number
+ */
+static void sigusr1_handler(int sig);
+
+int main(int argc, char *argv[]) {
+  openlog("aesdsocket", LOG_PID, LOG_USER);
+  syslog(LOG_DEBUG, "Starting `aesdsocket`.");
+
+  // Parse Arguments
+  bool execute_as_daemon = false;
+  if ((argc == 2) && (strcmp(argv[1], "-d") == 0)) {
+    execute_as_daemon = true;
+  } else if (argc != 1) {
+    printf("Usage: %s [-d]\r\n", argv[0]);
+    printf("Options:\r\n");
+    printf("  -d    execute as deamon\r\n");
+    closelog();
+    return 1;
+  }
+
+  int server_socket = 0;
+  struct addrinfo *server_addrinfo = NULL;
+  if (setup_socket_server(execute_as_daemon, &server_socket,
+                          &server_addrinfo)) {
+    syslog(LOG_ERR, "setup_socket_server");
+    close(server_socket);
+    freeaddrinfo(server_addrinfo);
+    closelog();
+    return -1;
+  }
+
+  timer_t log_timestamp_timer;
+  if (setup_signal_handlers(&log_timestamp_timer)) {
+    syslog(LOG_ERR, "setup_signal_handlers");
+    timer_delete(log_timestamp_timer);
+    close(server_socket);
+    freeaddrinfo(server_addrinfo);
+    closelog();
+    return -1;
+  }
+
+  // Setup Timestamp Logger
+  pthread_t timestamp_thread_id;
+  const int pthread_create_result =
+      pthread_create(&timestamp_thread_id, NULL, log_timestamp_worker, NULL);
+  if (pthread_create_result) {
+    syslog(LOG_ERR, "pthread_create returned error: %d", pthread_create_result);
+    timer_delete(log_timestamp_timer);
+    close(server_socket);
+    freeaddrinfo(server_addrinfo);
+    closelog();
+    return -1;
+  }
+
+  // Initialize result file mutex
+  if (pthread_mutex_init(config_get_result_file_mutex(), NULL)) {
```

Comment:
Based on the class conversation, I should initialize this mutex before the timer.